k4FWCore/include/k4FWCore/FunctionalUtils.h

```diff
@@ -142,16 +142,50 @@ namespace details {
         using EDM4hepType = std::remove_cv_t<std::remove_pointer_t<typename TupleType::value_type>>;
         auto inputMap = std::vector<const EDM4hepType*>();
         for (auto& handle : std::get<Index>(handles)) {
-          podio::CollectionBase* in = handle.get()->get();
-          auto* typedIn = dynamic_cast<const EDM4hepType*>(in);
-          if (typedIn) {
-            inputMap.push_back(typedIn);
-          } else {
-            throw GaudiException(
-                thisClass->name(),
-                fmt::format("Failed to cast collection {} to the required type {}, the type of the collection is {}",
-                            handle.objKey(), typeid(EDM4hepType).name(), in ? in->getTypeName() : "[undetermined]"),
-                StatusCode::FAILURE);
+          try {
+            podio::CollectionBase* in = handle.get()->get();
+            auto* typedIn = dynamic_cast<const EDM4hepType*>(in);
+            if (typedIn) {
+              inputMap.push_back(typedIn);
+            } else {
+              throw GaudiException(
+                  thisClass->name(),
+                  fmt::format("Failed to cast collection {} to the required type {}, the type of the collection is {}",
+                              handle.objKey(), typeid(EDM4hepType).name(), in ? in->getTypeName() : "[undetermined]"),
+                  StatusCode::FAILURE);
+            }
+          } catch (GaudiException& e) {
+            // When the type of the collection is different from the one requested, this can happen because
+            // 1. a mistake was made in the input types of a functional algorithm
+            // 2. the data was produced using the old DataHandle, which is never going to be in the input type
+            if (e.message().find("different from") != std::string::npos) {
```

Comment:
Is it possible to avoid using an exception and message checking here for control flow?

Reply:
I guess one could change how stuff is stored by the PodioDataSvc but I'm not sure it's worth it. I don't think it takes that much time since this can happen often but not that often (once per event per collection produced by an old algorithm and per algorithm consuming that collection).